doc/manpages/katprep_authconfig.1.md
% katprep_authconfig(1) Version 0.5.0 | katprep documentation

NAME
====

**katprep_authconfig** — manages credentials for third-party systems triggered by katprep

SYNOPSIS
========

| **katprep_authconfig** \[**-h**] \[**-v**] \[**-q**] \[**-d**] \[_file_] \[**list**|**add**|**remove**|**password**]

DESCRIPTION
===========

Creates, modifies and removes entries from authentication containers used by the katprep framework in order to gain access to external third-party systems (such as monitoring systems and hypervisors). This removes the need of entering login information every time to trigger external systems.
Authentication containers are JSON documents that can also be protected by a passphrase. In this case, you need to enter the passphrase once when using the container.

Options
-------

-h, --help

:   Prints brief usage information.

-v, --version

:   Prints the current version number.

-q, --quiet

:   Supresses printing status messages to stdout.

-d, --debug

:   Enables debugging outputs.

Listing credentials
-------------------

To list credentials, use the **list** command. By default, the output will contain hostnames and usernames, but no password. To also show password in plain text, add the following parameter:

-a, --show-password
:    also print passwords.

Adding credentials
------------------

To add credentials, use the **add** command. By default, you will be prompted for hostname, username and password. To pre-select information, utilize the following parameters:

-H _hostname_, --hostname _hostname_

:   Third-party system hostname

-u _username_, --username _username_

:   Appropriate username

-p _password_, --password _password_

:   Corresponding password

Removing credentials
--------------------

To remove credentials, use the **remove** command. You will be prompted for a hostname, to pre-select the hostname, utilize the following parameter:

-H _hostname_, --hostname _hostname_

:   Third-party system hostname

Encrypting/decrypting containers
--------------------------------

By default, authentication containers contain login information in plain text. To enhance security, it is possible to encrypt the passwords with a passphrase up to 32 chars. To encrypt or decrypt a file, utilize the **password** command. By default, the utility prompts a password. To pre-select the password, utilize the following parameter:

-p _password_, --password _password_

:   Password

To encrypt an authentication container, simply execute **katprep_authconfig** \[_file_] **password** and specify a passphrase. To remove the encryption, re-run the command without specifying a passphrase. 

FILES
=====

*~/.katpreprc*

:   Per-user katprep configuration file.

*katprep.auth*

:   Individual katprep authentication container file.

BUGS
====

See GitHub issues: <https://github.com/stdevel/katprep/issues>

AUTHOR
======

Christian Stankowic <info@cstan.io>

SEE ALSO
========

**katprep.auth(5)**


katprep_authconfig.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
A script which maintains entries in a authentication container.
"""

import argparse
import logging
import json
import getpass
from AuthContainer import AuthContainer

__version__ = "0.0.1"
LOGGER = logging.getLogger('katprep_authconfig')
"""
logging: Logger instance
"""
CONTAINER = None
"""
AuthContainer: authentication container file
"""



def list(args):
    """
    This function lists entries from the authentication container.
    """
    for hostname in CONTAINER.get_hostnames():
        #get credentials
        credentials = CONTAINER.get_credential(hostname)
        if options.show_passwords:
            password = credentials[1]
        else:
            password = "xxx"
        #print entry
        print "{} (Username: {} / Password: {})".format(
            hostname, credentials[0], password
        )



def add(args):
    """
    This function adds/modifies an entry to/from the authentication container.
    """
    while options.entry_hostname == "":
        #prompt for hostname
        options.entry_hostname = raw_input("Hostname: ")
    while options.entry_username == "":
        #prompt for hostname
        options.entry_username = raw_input(
            "{} Username: ".format(options.entry_hostname)
        )
    while options.entry_password == "":
        #prompt for password
        options.entry_password = getpass.getpass(
            "{} Password: ".format(options.entry_hostname)
        )
    #prompt again
    verification = ""
    while verification != options.entry_password:
        verification = getpass.getpass(
            "Verify {} Password: ".format(options.entry_hostname)
        )
    LOGGER.debug("Adding entry hostname='{}', username='{}'...".format(
        options.entry_hostname, options.entry_username)
    )
    CONTAINER.add_credentials(
        options.entry_hostname, options.entry_username, options.entry_password
    )
    CONTAINER.save()



def remove(args):
    """
    This function removes an entry from the authentication container.
    """
    while options.entry_hostname == "":
        #prompt for hostname
        options.entry_hostname = raw_input("Hostname: ")
    LOGGER.debug("Removing entry hostname='{}'...".format(
        options.entry_hostname)
    )
    CONTAINER.remove_credentials(options.entry_hostname)
    CONTAINER.save()



def parse_options(args=None):
    """Parses options and arguments."""
    desc = '''katprep_authconfig.py is used for creating, modifying and
    removing entries in/from an authentication container.
    Authentication containers include various authentication credentials for
    external systems that can be accessed from the katprep utilities (e.g.
    monitoring systems, hypervisor connections, etc.).
    This will make system maintenance automation easier as you don't have to
    enter credentials every time.
    '''
    epilog = '''Check-out the website for more details:
     http://github.com/stdevel/katprep'''
    parser = argparse.ArgumentParser(
        description=desc, version=__version__, epilog=epilog
    )

    #define option groups
    gen_opts = parser.add_argument_group("generic arguments")

    #GENERIC ARGUMENTS
    #-q / --quiet
    gen_opts.add_argument("-q", "--quiet", action="store_true", \
    dest="generic_quiet", \
    default=False, help="don't print status messages to stdout (default: no)")
    #-d / --debug
    gen_opts.add_argument("-d", "--debug", dest="generic_debug", \
    default=False, action="store_true", \
    help="enable debugging outputs (default: no)")
    #authentication container
    gen_opts.add_argument('container', metavar='FILE', nargs=1, \
    help='An authentication container', type=str)

    #COMMANDS
    subparsers = parser.add_subparsers(title='commands', \
    description='controlling maintenance stages', help='additional help')
    cmd_list = subparsers.add_parser("list", help="listing entries")
    cmd_list.add_argument("-a", "--show-passwords", action="store_true", \
    dest="show_passwords", default=False, help="also shows passwords " \
    "(default: no)")
    cmd_list.set_defaults(func=list)

    cmd_add = subparsers.add_parser("add", help="adding/modifying entries")
    cmd_add.set_defaults(func=add)
    cmd_add.add_argument("-H", "--hostname", action="store", default="", \
    dest="entry_hostname", metavar="HOSTNAME", help="hostname entry")
    cmd_add.add_argument("-u", "--username", action="store", default="", \
    dest="entry_username", metavar="USERNAME", help="username")
    cmd_add.add_argument("-p", "--password", action="store", default="", \
    dest="entry_password", metavar="PASSWORD", help="corresponding password")

    cmd_remove = subparsers.add_parser("remove", help="removing entries")
    cmd_remove.add_argument("-H", "--hostname", action="store", default="", \
    dest="entry_hostname", metavar="HOSTNAME", help="hostname entry")
    cmd_remove.set_defaults(func=remove)

    #parse options and arguments
    options = parser.parse_args()
    return (options, args)



def main(options, args):
    """Main function, starts the logic based on parameters."""
    global CONTAINER

    LOGGER.debug("Options: {0}".format(options))
    LOGGER.debug("Arguments: {0}".format(args))

    #load container
    CONTAINER = AuthContainer(options.container[0])

    #start action
    options.func(options.func)



if __name__ == "__main__":
    (options, args) = parse_options()

    #set logging level
    logging.basicConfig()
    if options.generic_debug:
        LOGGER.setLevel(logging.DEBUG)
    elif options.generic_quiet:
        LOGGER.setLevel(logging.ERROR)
    else:
        LOGGER.setLevel(logging.INFO)

    main(options, args)
